contracts/src/v0.8/dev/VRFCoordinatorV2.sol

```diff
@@ -0,0 +1,716 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "../interfaces/LinkTokenInterface.sol";
+import "../interfaces/BlockhashStoreInterface.sol";
+import "../interfaces/AggregatorV3Interface.sol";
+import "../interfaces/TypeAndVersionInterface.sol";
+
+import "./VRF.sol";
+import "../ConfirmedOwner.sol";
+import "../interfaces/VRFConsumerV2Interface.sol";
+
+contract VRFCoordinatorV2 is VRF, ConfirmedOwner, TypeAndVersionInterface {
+
+  LinkTokenInterface public immutable LINK;
+  AggregatorV3Interface public immutable LINK_ETH_FEED;
+  BlockhashStoreInterface public immutable BLOCKHASH_STORE;
+
+  error InsufficientBalance();
+  error InvalidConsumer(uint64 subId, address consumer);
+  error InvalidSubscription();
+  error AlreadySubscribed(uint64 subId, address consumer);
+  error OnlyCallableFromLink();
+  error InvalidCalldata();
+  error MustBeSubOwner(address owner);
+  error MustBeRequestedOwner(address proposedOwner);
+  error BalanceInvariantViolated(uint256 internalBalance, uint256 externalBalance); // Should never happen
+  event FundsRecovered(address to, uint256 amount);
+  // There are only 1e9*1e18 = 1e27 juels in existence, so the balance can fit in uint96 (2^96 ~ 7e28)
+  struct Subscription {
+    uint96 balance; // Common link balance used for all consumer requests.
+    address owner; // Owner can fund/withdraw/cancel the sub.
+    address requestedOwner; // For safe transfering sub ownership.
+  }
+  // We make this public so people can lookup which subscription a given consumer is using (if any).
+  mapping(address /* consumer */ => uint64 /* subId */) public s_consumers;
+  mapping(uint64 /* subId */ => Subscription /* subscription */) private s_subscriptions;
+  uint64 private s_currentSubId;
+  // s_totalBalance tracks the total link sent to/from
+  // this contract through onTokenTransfer, defundSubscription, cancelSubscription and oracleWithdraw.
+  // A discrepancy with this contracts link balance indicates someone
+  // sent tokens using transfer and so we may need to use recoverFunds.
+  uint96 public s_totalBalance;
+  event SubscriptionCreated(uint64 indexed subId, address owner, address[] consumers);
+  event SubscriptionFunded(uint64 indexed subId, uint256 oldBalance, uint256 newBalance);
+  event SubscriptionConsumerAdded(uint64 indexed subId, address consumer);
```

Comment:
In order to keep gas costs low in requestRandomWords and avoid having a user specify a consumerID (index into an array of consumers) I used the ```s_consumers``` map. One consequence of that is it makes it more difficult to answer the question "give me all the addresses associate with this subscription". We could maintain an auxiliary data structures which has subId=>[consumerAddress1, consumerAddress2....], specifically to make answering that question easier. However I left that out for a few reasons:
1. We'd have to have a maxConsumers limit also as we would need to loop over subId in add/removeConsumer
2. People can use the s_consumers map directly to answer individual subscription questions like "what subscription is X using". In practice I would imagine this is the dominant use case. Sub owners would know which addresses they shared their subID with.
3. As a backup you could easily compute the consumer set using the logs SubscriptionCreated, SubscriptionConsumerAdded, SubscriptionConsumerRemoved